R/implicit_assignment_linter.R

```diff
@@ -0,0 +1,112 @@
+#' Avoid implicit assignment in function calls
+#'
+#' Assigning inside function calls makes the code difficult to read, and should
+#' be avoided, except for functions that capture side-effects (e.g. [capture.output()]).
+#'
+#' @param except A character vector of functions to be excluded from linting.
+#'
+#' @examples
+#' # will produce lints
+#' lint(
+#'   text = "if (x <- 1L) TRUE",
+#'   linters = implicit_assignment_linter()
+#' )
+#'
+#' lint(
+#'   text = "mean(x <- 1:4)",
+#'   linters = implicit_assignment_linter()
+#' )
+#'
+#' # okay
+#' writeLines("x <- 1L\nif (x) TRUE")
+#' lint(
+#'   text = "x <- 1L\nif (x) TRUE",
+#'   linters = implicit_assignment_linter()
+#' )
+#'
+#' writeLines("x <- 1:4\nmean(x)")
+#' lint(
+#'   text = "x <- 1:4\nmean(x)",
+#'   linters = implicit_assignment_linter()
+#' )
+#'
+#' @evalRd rd_tags("implicit_assignment_linter")
+#' @seealso
+#' - [linters] for a complete list of linters available in lintr.
+#' - <https://style.tidyverse.org/syntax.html#assignment>
+#'
+#' @export
+implicit_assignment_linter <- function(except = c(
+                                         "bquote", "capture.output", "expression", "expr",
+                                         "expect_error", "expect_warning", "expect_message",
+                                         "expect_no_error", "expect_no_warning", "expect_no_message",
+                                         "expect_condition", "expect_no_condition",
+                                         "expect_invisible", "expect_visible",
+                                         "expect_output", "expect_silent",
+                                         "local", "quo", "quos", "quote", "test_that"
+                                       )) {
```

Comment:
@AshesITR I styled the file with `{styler}`, but still can't get rid of the lint from `indentation_linter`:

> Line 40: Indentation should be 2 spaces but is 41 spaces. 

So the builds are going to fail for now.

Reply:
There is a line length lint as well.
styler seems off here, this is similar to a multiline function call, which should be block-indented.

Compare
```r
lapply(x, FUN = function(i) {
  42L
})
```

Lintr expects this to look like
```r
implicit_assignment_linter <- function(except = c(
  "bquote", "capture.output", "expression", "expr",
  "expect_error", "expect_warning", "expect_message",
  "expect_no_error", "expect_no_warning", "expect_no_message",
  "expect_condition", "expect_no_condition",
  "expect_invisible", "expect_visible",
  "expect_output", "expect_silent",
  "local", "quo", "quos", "quote", "test_that"
)) { 
  # code ... 
}
```